client/consul/sync.go

```diff
@@ -0,0 +1,342 @@
+			if err := c.registerService(srv); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+			}
+		}
+		c.services[srv.ID] = srv
+		services[srv.ID] = srv
+
+		for _, chk := range service.Checks {
+			checkID := chk.Hash(srv.ID)
+			if _, ok := c.checks[checkID]; !ok {
+				if err := c.registerCheck(chk, srv); err != nil {
+					mErr.Errors = append(mErr.Errors, err)
+				}
+			}
+			c.checks[checkID] = chk
+			checks[checkID] = chk
+		}
+	}
+
+	// Remove services that are not present anymore
+	for _, service := range c.services {
+		if _, ok := services[service.ID]; !ok {
+			if err := c.deregisterService(service.ID); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+			}
+			delete(c.services, service.ID)
+		}
+	}
+
+	// Remove the checks that are not present anymore
+	for checkID, _ := range c.checks {
+		if _, ok := checks[checkID]; !ok {
+			if err := c.deregisterCheck(checkID); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+			}
+			delete(c.checks, checkID)
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+// Shutdown de-registers the services and checks and shuts down periodic syncing
+func (c *ConsulService) Shutdown() error {
+	var mErr multierror.Error
+	select {
+	case c.shutdownCh <- struct{}{}:
+	default:
+	}
+	for _, service := range c.services {
+		if err := c.client.Agent().ServiceDeregister(service.ID); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+func (c *ConsulService) RemoveServices(tasks []*structs.Task) error {
+	var mErr multierror.Error
+	var services map[string]struct{}
+	for _, task := range tasks {
+		for _, service := range task.Services {
+			services[service.ID()] = struct{}{}
+		}
+	}
+
+	cServices, err := c.client.Agent().Services()
+	if err != nil {
+		return err
+	}
+	cServices = c.filterConsulServices(cServices)
+
+	for _, service := range cServices {
+		if _, validService := services[service.ID]; !validService {
+			if err := c.deregisterService(service.ID); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+			}
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+// registerCheck registers a check definition with Consul
+func (c *ConsulService) registerCheck(check *structs.ServiceCheck, service *consul.AgentService) error {
+	chkReg := consul.AgentCheckRegistration{
+		ID:        check.Hash(service.ID),
+		Name:      check.Name,
+		ServiceID: service.ID,
+	}
+	chkReg.Timeout = check.Timeout.String()
+	chkReg.Interval = check.Interval.String()
+	switch check.Type {
+	case structs.ServiceCheckHTTP:
+		if check.Protocol == "" {
+			check.Protocol = "http"
+		}
+		url := url.URL{
+			Scheme: check.Protocol,
+			Host:   fmt.Sprintf("%s:%d", service.Address, service.Port),
+			Path:   check.Path,
+		}
+		chkReg.HTTP = url.String()
+	case structs.ServiceCheckTCP:
+		chkReg.TCP = fmt.Sprintf("%s:%d", service.Address, service.Port)
+	case structs.ServiceCheckScript:
+		chkReg.TTL = check.Interval.String()
```

Comment:
@dadgar We could ask consul to do it, but since we are not passig the driver name in the context I was thinking to do this in the executor as well.